### PR TITLE
Asyncify filter middleware

### PR DESCRIPTION
**Pull request by pacrob**
### What was wrong?

Filter middleware needed asyncing

### How was it fixed?

asynced

#### Cute Animal Picture

image

### CODE REVIEW

**Comment by fselmo** [comment]
👀 I'll take a peek here this week!

File: tests/core/middleware/test_filter_middleware.py [hunk 1]
```diff
@@ -197,6 +247,82 @@ def test_iter_latest_block_ranges(
...
+)
+@pytest.mark.asyncio
+async def test_async_iter_latest_block_ranges(
+    async_w3, iter_block_number, from_block, to_block, current_block, expected
+):
+    latest_block_ranges = async_iter_latest_block_ranges(async_w3, from_block, to_block)
+    for index, block in enumerate(current_block):
+        iter_block_number.send(block)
+        expected_tuple = expected[index]
+        actual_tuple = await latest_block_ranges.__anext__()
```
[the remark this answers is not in the document]
Comment (pacrob) [reply tests/core/middleware/test_filter_middleware.py:322]: `anext` is only available as of [3.10](https://docs.python.org/3/whatsnew/3.10.html#other-language-changes)

[the remark this answers is not in the document]
Comment (fselmo) [reply tests/core/middleware/test_filter_middleware.py:322]: Yep, nice catch 👌🏼

File: web3/middleware/filter.py [hunk 2]
```diff
@@ -293,6 +392,94 @@ def get_logs(self) -> List[LogReceipt]:
...
+    async def _get_filter_changes(self) -> AsyncIterator[List[LogReceipt]]:
+        self_from_block = await self.from_block
+        self_to_block = await self.to_block
+        async for start, stop in async_iter_latest_block_ranges(
+            self.w3, self_from_block, self_to_block
+        ):
+            if None in (start, stop):
+                yield []
+            else:
+                yield [
```
[the remark this answers is not in the document]
Comment (pacrob) [reply web3/middleware/filter.py:545]: fixed

File: tests/core/middleware/test_filter_middleware.py [hunk 3]
```diff
@@ -85,6 +100,18 @@ def result_generator_middleware(iter_block_number):
     )
 
 
+@pytest_asyncio.fixture(scope="function")
```
[the remark this answers is not in the document]
Comment (pacrob) [reply tests/core/middleware/test_filter_middleware.py]: done!

File: web3/middleware/filter.py [hunk 4]
```diff
@@ -158,6 +163,42 @@ def iter_latest_block(
         _last = latest_block
 
 
+async def async_iter_latest_block(
```
[the remark this answers is not in the document]
Comment (pacrob) [reply web3/middleware/filter.py]: done!

**Pull request merged**